include/exec/sequence/transform_each.hpp

```diff
@@ -0,0 +1,161 @@
+            && __nothrow_callable<_Adaptor&, _Item>)
+          -> __next_sender_of_t<_Receiver, __call_result_t<_Adaptor&, _Item>> {
+          return exec::set_next(
+            __self.__op_->__receiver_, __self.__op_->__adaptor_(static_cast<_Item&&>(__item)));
+        }
+
+        template <same_as<set_value_t> _SetValue, same_as<__t> _Self>
+        friend void tag_invoke(_SetValue, _Self&& __self) noexcept {
+          stdexec::set_value(static_cast<_Receiver&&>(__self.__op_->__receiver_));
+        }
+
+        template <same_as<set_stopped_t> _SetStopped, same_as<__t> _Self>
+          requires __callable<_SetStopped, _Receiver&&>
+        friend void tag_invoke(_SetStopped, _Self&& __self) noexcept {
+          stdexec::set_stopped(static_cast<_Receiver&&>(__self.__op_->__receiver_));
+        }
+
+        template <same_as<set_error_t> _SetError, same_as<__t> _Self, class _Error>
+          requires __callable<_SetError, _Receiver&&, _Error>
+        friend void tag_invoke(_SetError, _Self&& __self, _Error&& __error) noexcept {
+          stdexec::set_error(
+            static_cast<_Receiver&&>(__self.__op_->__receiver_), static_cast<_Error&&>(__error));
+        }
+
+        template <same_as<get_env_t> _GetEnv, __decays_to<__t> _Self>
+        friend env_of_t<_Receiver> tag_invoke(_GetEnv, _Self&& __self) noexcept {
+          return stdexec::get_env(__self.__op_->__receiver_);
+        }
+      };
+    };
+
+    template <class _Sender, class _ReceiverId, class _Adaptor>
+    struct __operation {
+      using _Receiver = stdexec::__t<_ReceiverId>;
+
+      struct __t : __operation_base<_Receiver, _Adaptor> {
+        subscribe_result_t<_Sender, stdexec::__t<__receiver<_ReceiverId, _Adaptor>>> __op_;
+
+        __t(_Sender&& __sndr, _Receiver __rcvr, _Adaptor __adaptor)
+          : __operation_base<
+            _Receiver,
+            _Adaptor>{static_cast<_Receiver&&>(__rcvr), static_cast<_Adaptor&&>(__adaptor)}
+          , __op_{exec::subscribe(
+              static_cast<_Sender&&>(__sndr),
+              stdexec::__t<__receiver<_Receiver, _Adaptor>>{this})} {
+        }
+
+        friend void tag_invoke(start_t, __t& __self) noexcept {
+          stdexec::start(__self.__op_);
+        }
+      };
+    };
+
+    template <class _SenderId, class _Adaptor>
+    struct __sender {
+      using _Sender = stdexec::__t<_SenderId>;
+
+      struct __t {
+        using is_sender = sequence_tag;
+        using __id = __sender;
+
+        _Sender __sender_;
+        _Adaptor __adaptor_;
+
+        template <class _Self, class _Env>
+        using __some_sender = __copy_cvref_t<
+          _Self,
+          __sequence_sndr::__unspecified_sender_of<
+            completion_signatures_of_t<__copy_cvref_t<_Self, _Sender>, _Env>>>;
+
+        template <class _Self, class _Env>
+        using completion_sigs_t = completion_signatures_of_t<
+          __call_result_t<_Adaptor&, __some_sender<_Self, _Sender>>,
```

Comment:
Is this supposed to be:

```suggestion
          __call_result_t<_Adaptor&, __some_sender<_Self, _Env>>,
```

? Also, I don't think I really understand what's going on here. We are saying the completion signatures of `transform_each` are the completions of a sender type that is returned from the adaptor when passed ... some unspecified sender type? Doesn't it matter which sender is passed to the adaptor?

Reply:
We pass the items of the predecessor sequence to the adaptor. In the current design we chose to not list their concrete item type somewhere. I think it's inaccessible in this context here. One could imagine a design where a sequence does not specify completion signatures but it's item sender types instead.

You will find a similar use of an unspecified sender in the sequence sender concepts for this reason.

The choice to specify only completion signatures of items  has been made for convenience. But from this follows some consistency assumptions, i. e. If I pass an item with signatures S I always get an item with signatures R, indepently of the concrete item type.

A similar consistency assumption is present for set next.

Reply:
> from this follows some consistency assumptions, i. e. If I pass an item with signatures S I always get an item with signatures R, indepently of the concrete item type.

This makes me a little uncomfortable. Algorithms are customizable, and I can imagine some implementations might change signatures slightly: an additional error, or a stop signal.

Reply:
I have addressed that now with PR #1020 which is included here.